chacha20_poly1305/src/chacha20.rs

```diff
@@ -0,0 +1,460 @@
+// SPDX-License-Identifier: CC0-1.0
+
+//! The ChaCha20 stream cipher from RFC8439.
+
+use core::ops::BitXor;
+
+/// The first four words (32-bit) of the ChaCha stream cipher state are constants.
+const WORD_1: u32 = 0x61707865;
+const WORD_2: u32 = 0x3320646e;
+const WORD_3: u32 = 0x79622d32;
+const WORD_4: u32 = 0x6b206574;
+
+/// The cipher's block size is 64 bytes.
+const CHACHA_BLOCKSIZE: usize = 64;
+
+/// A 256-bit secret key shared by the parties communicating.
+#[derive(Clone, Copy)]
+pub struct Key([u8; 32]);
+
+impl Key {
+    /// Create a new key.
+    pub const fn new(key: [u8; 32]) -> Self {
+        Key(key)
+    }
+}
+
+/// A 96-bit initialization vector (IV), or nonce.
+#[derive(Clone, Copy, Debug)]
+pub struct Nonce([u8; 12]);
+
+impl Nonce {
+    /// Create a new nonce.
+    pub const fn new(nonce: [u8; 12]) -> Self {
+        Nonce(nonce)
+    }
+}
+
+/// A SIMD-friendly structure which holds 25% of the cipher state.
+///
+/// The cipher's quarter round function is the bulk of its work
+/// and there are large performance gains to be had if the function
+/// leverages SIMD instructions on architectures which support them. Because
+/// the algorithm allows for the cipher's state to be operated on in
+/// parallel (each round only touches a quarter of the state), then theoretically
+/// the parallel SIMD instructions should be used. But sometimes the
+/// compiler needs a few hints to ensure it recognizes a "vectorizable" function.
+/// That is the goal of this type, which clearly breaks the state up into four
+/// chunks and exposes functions which align with SIMD lanes.
+///
+/// This type is attempting to be as close as possible to the experimental [`core::simd::u32x4`]
+/// which at this time is feature gated and well beyond the project's MSRV. But ideally
+/// an easy transistion can be made in the future.
+///
+/// A few SIMD relevant design choices:
+///    * Heavy use of inline function calls.
+///    * For-each loops are easy for the compiler to recognize as vectorizable.
+///    * The type is a based on an array instead of tuple since the heterogeneous
+///      nature of tuples can confuse the compiler into thinking it is not vectorizable.
+///    * Memory layout lines up with SIMD size.
+#[repr(C, align(16))]
```

Comment:
Why C and not `transparent`? (I know nothing about the topic, so I'm just curious.)

Also could we disable the alignment on architectures that definitely don't have SIMD?

Reply:
I don't have a ton of experience here either, but after reading a bit on the docs I don't think there would be a huge difference between `C` and `transparent` in this case. `C` is the most explicit "16 bytes", `transparent` is less explicit but I think will almost always be 16 bytes as well since it is based on the `[u32; 4]`. The `C` version perhaps is the most clear on what we are asking for here?

I am guessing we could do something like `#[cfg_attr(any(target_arch = "x86", target_arch = "x86_64", target_arch = "aarch64"), repr(C, align(16)))]` (haven't tested this) to only enable on certain architecture? Kinda gets at the heart of the SIMD tradeoff where readability/maintenance might be a burden. I don't have a strong feeling on this one. The struct is relatively small, but used quite often, so not sure how big an issue wasted memory would be on non-SIMD architectures.

Reply:
Oh, transparent would be useless because we're not doing any `unsafe` casts. However it does look weird to me that `C` would actually affect codegen and it looks confusing. Doesn't it work without the `C`?

And yes, I was thinking of some `cfg_attr` but not sure if it should be whitelist, blacklist or based on `target_feature`. And I think the architectures that don't support SIMD tend to have other limits such as less RAM so not wasting the space could be important. (Certainly something that can be improved in future PRs.)

Reply:
To be clear, you don't need to blacklist AVR in this PR.